README.md
# Rabbitr
[![TravisCI](https://img.shields.io/travis/urbanmassage/node-rabbitr.svg)](https://travis-ci.org/urbanmassage/node-rabbitr)
[![npm](https://img.shields.io/npm/v/rabbitr.svg)](https://www.npmjs.com/package/rabbitr)
[![npm](https://img.shields.io/npm/dt/rabbitr.svg)](https://www.npmjs.com/package/rabbitr)
[![Codecov](https://img.shields.io/codecov/c/github/urbanmassage/node-rabbitr.svg)](https://codecov.io/github/urbanmassage/node-rabbitr)

RabbitMQ made easy for nodejs

## Init / setup

```js
var rabbit = new Rabbitr({
	host: 'localhost',
});
```

## Basic queue usage

```js
// in one module
rabbit.subscribe(['booking.create'], 'sms.send.booking.create', {}, (message) => {
	// send an sms
	message.ack();
});

// in another module
rabbit.subscribe(['booking.create'], 'email.send.booking.create', {}, (message) => {
	// send an email
	message.ack();
});

// elsewhere
rabbit.send('booking.create', {id: 1});
```

## Timers
Rabbitr makes using dead letter exchanges dead easy

```js
// set timer
rabbit.subscribe(['booking.create'], 'booking.not-confirmed.timer.set', {}, (message) => {
  // do something to calculate how long we want the timer to last
	var timeFromNow = 900000; // 15 mins

	rabbit.setTimer('booking.not-confirmed.timer.fire', message.data.id, {
    id: message.data.id,
	}, timeFromNow);

	message.ack();
});

// clear timer if something has happened that means the timer action isn't required
rabbit.subscribe(['booking.confirm'], 'booking.not-confirmed.timer.clear', {}, (message) => {
	rabbit.clearTimer('booking.not-confirmed.timer.fire', message.data.id);

	message.ack();
});

// handle the timer firing
rabbit.subscribe(['booking.not-confirmed.timer.fire'], 'booking.not-confirmed.timer.fire', {}, (message) => {
	// do something off the back of the timer firing
	// in this example, message.data.id is the booking id that wasn't confirmed in time
	message.ack();
});
```

## RPC (remote procedure call)
Use Rabbitr's RPC methods if you need to do something and get a response back, and you want to decouple the two processes via MQ

- Make sure you use the same major version of Rabbitr on both the worker and scheduler sides!

### Define the worker's method
Use `prefetch` in the options object to define concurrency (defaults to `1`).

```js
rabbit.rpcListener('rpc-test', { prefetch: 5 }, async (message) => {
	// do something with message.data

	await doSomethingAsyncThatMightThrow(message.data);

  return {
    rpc: 'is cool'
	};
});
```

### Calling the worker's RPC
Define the timeout in milliseconds in the options object for `rpcExec`

```js
rabbit.rpcExec('rpc-test', { some: 'data' }, { timeout: 5000 }).then((response) => {
	// do something with `response`
	// it will look like { rpc: 'is cool' }
});
```

## Backoff
Rabbitr supports a few backoff types
- Immediate
- Interval
- Exponential

These backoff types can either be set globally against the whole of rabbitr by passing it into your initialization

```js
rabbit = new Rabbitr({
  url: 'amqp://guest:guest@localhost/%2F',
  backoffLogic: backoff
})
```
or can be set at the individual subscriber level
```js
rabbit.subscribe([exchangeName], queueName, {backoffLogic:backoff}, (message) => {
  // Your logic goes here
});
```

If there is no logic set at the subscriber level then we will fall back to the global backoff (if one is configured)

### Backoff types
Here is a description of the currently implemented backoffs
#### Immediate
This logic is used if you want to retry a certain amount of times without waiting, this can initialized with the following `new Immmediate({number of times to retry})`
#### Interval
Much like the Immediate logic this will retry a set amount of times, but instead you pass the intervals that you want to retry with `new Interval([{values for retry}])`
#### Exponential
This uses a max number of retries and a multiplier to determine the amount of seconds to wait `new Exponential({multiplier}, {number of times to retry})`

## Debugging

To debug rabbitr, you can enable logging by setting the environment variable
`DEBUG` to "rabbitr".

You can also tell rabbitr to only listen on one or few queues using the
environment variable `RABBITR_DEBUG`. Just set it to a comma-separated list of
queues names. RPC queues have the prefix `rpc.`.

```bash
# To enable logging
DEBUG=rabbitr node .

# To only listen on rpc channels `user.create` and `user.update`
RABBITR_DEBUG=rpc.user.create,rpc.user.update node .
```


CHANGELOG.md
# CHANGELOG

# 12.1.0
- Adds support for `context` within the opts object of `rpcExec` - this then appears in `message.context` on the `rpcListener` side

# 12.0.0
- Removes support for `queuePrefix` in the constructor options
- Updates amqplib to `0.7.1`
- Moves away from deprecated `new Buffer()` constructor format to `Buffer.from()` internally

## 11.1.9
- Build on node 10 + 12

## 11.1.8
- Handle a null / undefined `opts` parameter for the `subscribe` method without getting into an error state on rejection

## 11.0.0
- Removes support for node before version 6
- Major rewrite which changes a lot of interfaces
- Removed support for callbacks
- Made the options object required for a number of calls
- Removed `bindExchangeToQueue`, instead you should pass a list of exchanges to the `subscribe` method as the first argument
- Automatically adds a 2 second delay when you reject a message (unless `skipBackoff` passed in the `subscribe` options), adding a faux backoff mechanism

## 8.6.0
- graceful shutdown now triggered by SIGINT or SIGTERM

## 8.5.0
- Fixed a race condition that resulted in a fatal error `404, "NOT_FOUND - no queue 'x' in vhost '/'"` when the main thread on the client is blocked.

## 8.4.0
- fixed a bug that caused the return value of `rpcExec` and other methods to always be `undefined` if called before the connection was established.
- introduced debug mode (via `RABBITR_DEBUG` env variable. check README.md)
- graceful shutdown: when SIGINT is received rabbitr will stop listening for incoming
messages but will continue processing pending messages if any and then disconnect from rabbitmq.

## 8.3.0
- use node-uuid instead of shortid

## 8.2.3
- critical bugfix in `subscribe`/`rpcListener` and durable queues

## 8.2.1
- bugfix when calling any method before connection is established
- callback/promise in `rpcListener`

## 8.2.0
- promises support

## 8.1.0
- not publishing source `.ts` files to npm anymore
- using separate source maps instead of inline ones

## 8.0.8
- made `TimeoutError` more useful by adding the topic to the error message and fixing an issue with stack being missing.

## 8.0.6
- use `object-assign` instead of `merge`.

## 8.0.5
- output a typescript declaration file, reducing the number of typings dependencies.

## 8.0.0
 * All rpcListener queues will need re-creating during deployment in RabbitMQ
  if you ran an older version due to the move from durable to non-durable queues.


lib/rabbitr.js
var util = require('util');
var async = require('async');
var EventEmitter = require('events').EventEmitter;
var chalk = require('chalk');
var amqplib = require('amqplib/callback_api');
var debug = require('debug')('rabbitr');
var extend = require('util')._extend;
var shortId = require('shortid');

var kDefaultRPCExpiry = 15000; // 15 seconds

var lock = function(_, cb) { cb(function() { }); }

var Rabbitr = module.exports = function constructor(opts) {
    EventEmitter.call( this );
    
    // you MUST provide a 'url' rather than separate 'host', 'password', 'vhost' now
    var o = this.opts = {
        queuePrefix: '', // preffixed to all queue names - useful for environment and app names etc
        setup: function(done) { done(); }, // called once the connection is ready but before anything is bound (allows for ORM setup etc)
        connectionOpts: {
            heartbeat: 1
        },
        ackWarningTimeout: 5000,
        autoAckOnTimeout: null,
        defaultRPCExpiry: kDefaultRPCExpiry
    };
        
    for (var p in opts) {
        if (!opts.hasOwnProperty(p) || opts[p] == null) continue; // ignore null props on the opts object
        
        o[p] = opts[p];
    }
    
    this.ready = false;
    this.doneSetup = false;
    this.connected = false;
    this.subscribeQueue = [];
    this.bindingsQueue = [];
    this.sendQueue = [];
    this.setTimerQueue = [];
    this.clearTimerQueue = [];
    this.rpcListenerQueue = [];
    this.rpcExecQueue = [];

    this.middleware = [];
    
    var self = this;

    self._queueChannelCache = {};
    //self._exchangeChannelCache = {};
    self.subscribes = {}; // is this still necessary??

    // stored as a locally scoped var rather than as a prototype as we don't really want any body else calling this
    var connect = function connect() {
        debug('#connect');

        debug('using connection url', self.opts.url);

        amqplib.connect(self.opts.url, self.opts.connectionOpts, function(err, conn) {
            if(err) {
                throw err;
            }

            // make sure to close the connection if the process terminates
            process.once('SIGINT', conn.close.bind(conn));

            conn.on('close', function() {
                throw new Error('Disconnected from RabbitMQ');
            });
            /*conn.on('error', function(e) {
                console.log(e);
                throw new Error('Disconnected from RabbitMQ', e);
            });*/

            conn.createChannel(function(err, channel) {
                if(err) {
                    throw err;
                }

                self._timerChannel = channel;
                self._publishChannel = channel;
                self._cachedChannel = channel;

                // cache the connection and do all the setup work
                self.connection = conn;
                self.connected = true;

                var afterSetup = function() {
                    self.ready = true;

                    debug('ready and has queue sizes', self.subscribeQueue.length, self.bindingsQueue.length, self.rpcListenerQueue.length, self.sendQueue.length, self.rpcExecQueue.length);
                    for(var i=0; i<self.subscribeQueue.length; i++) {
                        self.subscribe(self.subscribeQueue[i].topic, self.subscribeQueue[i].opts, self.subscribeQueue[i].cb, true);
                    }
                    self.subscribeQueue = [];
                    for(var i=0; i<self.bindingsQueue.length; i++) {
                        self.bindExchangeToQueue(self.bindingsQueue[i].exchange, self.bindingsQueue[i].queue, self.bindingsQueue[i].cb, true);
                    }
                    self.bindingsQueue = [];
                    for(var i=0; i<self.rpcListenerQueue.length; i++) {
                        self.rpcListener(self.rpcListenerQueue[i].topic, self.rpcListenerQueue[i].opts, self.rpcListenerQueue[i].executor, true);
                    }
                    self.rpcListenerQueue = [];
                    
                    // send anything in send queue but clear it after
                    for(var i=0; i<self.sendQueue.length; i++) {
                        self.send(self.sendQueue[i].topic, self.sendQueue[i].data, self.sendQueue[i].cb, self.sendQueue[i].opts);
                    }
                    self.sendQueue = [];
                    
                    // send anything in setTimer queue but clear it after
                    for(var i=0; i<self.setTimerQueue.length; i++) {
                        self.setTimer(self.setTimerQueue[i].topic, self.setTimerQueue[i].uniqueID, self.setTimerQueue[i].data, self.setTimerQueue[i].ttl, self.setTimerQueue[i].cb);
                    }
                    self.setTimerQueue = [];
                    
                    // send anything in clearTimer queue but clear it after
                    for(var i=0; i<self.clearTimerQueue.length; i++) {
                        self.clearTimer(self.clearTimerQueue[i].topic, self.clearTimerQueue[i].uniqueID, self.clearTimerQueue[i].cb);
                    }
                    self.clearTimerQueue = [];
                    
                    // send anything in rpcExec queue but clear it after
                    for(var i=0; i<self.rpcExecQueue.length; i++) {
                        self.rpcExec(self.rpcExecQueue[i].topic, self.rpcExecQueue[i].data, self.rpcExecQueue[i].opts, self.rpcExecQueue[i].cb);
                    }
                    self.rpcExecQueue = [];
                };
                
                if(self.doneSetup) {
                    afterSetup();
                }
                else {
                    self.opts.setup(function(err) {
                        if(err) throw err;
                            
                        self.doneSetup = true;
                            
                        afterSetup();
                    });
                }
            });
        });
    };
    connect();
};
util.inherits(Rabbitr, EventEmitter);

Rabbitr.prototype._formatName = function _formatName(name) {
    var self = this;

    if(self.opts.queuePrefix != '') {
        name = self.opts.queuePrefix + '.' + name;
    }
    
    return name;
};


// standard pub/sub stuff
Rabbitr.prototype.send = function send(topic, data, cb, opts) {
    var self = this;
    
    if(!self.ready) {
        debug('adding item to send queue');
        return self.sendQueue.push({
            topic: topic,
            data: data,
            cb: cb,
            opts: opts
        });
    }
    
    debug(chalk.yellow('send'), topic, data, opts);

    self._publishChannel.assertExchange(self._formatName(topic), 'topic', {}, function() {
        self._publishChannel.publish(self._formatName(topic), '*', new Buffer(JSON.stringify(data)), {
            contentType: 'application/json'
        });

        if(cb) cb(null);
    });
};
Rabbitr.prototype.subscribe = function subscribe(topic, opts, cb, alreadyInQueue) {
    var self = this;
    
    if(!cb) {
        cb = opts;
        opts = null;
    }
    
    if(alreadyInQueue != true) {
        debug('adding item to sub queue');
        self.subscribeQueue.push({
            topic: topic,
            opts: opts,
            cb: cb
        });
    }
    
    if(!self.ready) {
        return;
    }
    
    debug(chalk.cyan('subscribe'), topic, opts);

    self.connection.createChannel(function(err, channel) {
        if(err) {
            self.emit('error', err);
            if(cb) cb(err);
            return;
        }

        channel.assertQueue(self._formatName(topic), {}, function(err, ok) {
            if(err) {
                return cb(err);
            }

            // should this be arbritrarily hard coded?
            channel.prefetch(opts ? opts.prefetch || 1 : 1);

            var processMessage = function processMessage(msg) {
                if(!msg) return;

                var data = msg.content.toString();
                if(msg.properties.contentType == 'application/json') {
                    data = JSON.parse(data);
                }

                debug('got', topic, data);
            
                var acked = false;

                var ackedTimer = setTimeout(function() {
                    self.emit('warning', {
                        type: 'ack.timeout',
                        queue: topic,
                        message: data
                    });

                    if(self.opts.autoAckOnTimeout === 'acknowledge') {
                        acked = true;
                        messageObject.acknowledge();
                    } 
                    else if(self.opts.autoAckOnTimeout === 'reject') {
                        acked = true;
                        messageObject.reject(true);
                    }
                }, self.opts.ackWarningTimeout);

                var message = {
                    topic: topic,
                    send: self.send.bind(self),
                    rpcExec: self.rpcExec.bind(self),
                    data: data,
                    channel: channel,
                    ack: function() {
                        debug('acknowledging message', topic, data);
                        
                        if(acked) return;
                        acked = true;
                        clearTimeout(ackedTimer);
                        
                        channel.ack(msg);
                    },
                    reject: function() {
                        debug('rejecting message', topic, data);
                        
                        if(acked) return;
                        acked = true;
                        clearTimeout(ackedTimer);
                        
                        channel.nack(msg);
                    }
                };

                var skipMiddleware = opts ? opts.skipMiddleware || false : false;

                if(skipMiddleware) {
                    return self.emit(topic, message);
                }

                self._runMiddleware(message, function(middlewareErr) {
                    // TODO - how to handle common error function thing for middleware?

                    self.emit(topic, message); 
                });
            };

            channel.consume(self._formatName(topic), processMessage, function(err, ok) {
                if(err) {
                    self.emit('error', err);
                    if(cb) cb(err);
                    return;
                }

                if(cb) cb(null);
            });
        });
    });
};
Rabbitr.prototype.bindExchangeToQueue = function bindExchangeToQueue(exchange, queue, cb, alreadyInQueue) {
    var self = this;

    if(alreadyInQueue != true) {
        debug('adding item to bindings queue');
        self.bindingsQueue.push({
            exchange: exchange,
            queue: queue,
            cb: cb
        });
    }

    if(!self.ready) {
        return;
    }
    
    debug(chalk.cyan('bindExchangeToQueue'), exchange, queue);

    self.connection.createChannel(function(err, channel) {
        if(err) { 
            self.emit('error', err);
            if(cb) cb(err);
            return;
        }

        channel.assertQueue(self._formatName(queue));
        channel.assertExchange(self._formatName(exchange), 'topic');

        channel.bindQueue(self._formatName(queue), self._formatName(exchange), '*', {}, function(err, ok) {
            if(err) {
                self.emit('error', err);
                if(cb) cb(err);
                return;
            }

            channel.close();

            if(cb) cb(null);
        });
    });
};


// timed queue stuff
Rabbitr.prototype._timerQueueName = function _timerQueueName(topic, uniqueID) {
    return 'dlq.' + topic + '.' + uniqueID;
};
Rabbitr.prototype.setTimer = function setTimer(topic, uniqueID, data, ttl, cb) {
    var self = this;
    
    if(!self.ready) {
        debug('adding item to setTimer queue');
        self.setTimerQueue.push({
            topic: topic,
            uniqueID: uniqueID,
            data: data,
            ttl: ttl,
            cb: cb
        });
        
        return;
    }

    var timerQueue = self._timerQueueName(topic, uniqueID);
    
    debug(chalk.yellow('setTimer'), topic, uniqueID, data);

    self._timerChannel.assertQueue(self._formatName(timerQueue), {
        durable: true,
        deadLetterExchange: self._formatName(topic),
        deadLetterRoutingKey: '*',
        expires: (ttl + 1000)
    }, function(err) {
        if(err) { 
            self.emit('error', err);
            if(cb) cb(err);
            return;
        }

        self._timerChannel.sendToQueue(self._formatName(timerQueue), new Buffer(JSON.stringify(data)), {
            contentType: 'application/json',
            expiration: ttl
        }, function(err) {
            if(err) { 
                self.emit('error', err);
                if(cb) cb(err);
                return;
            }

            if(cb) cb(null);
        });
    });
};
Rabbitr.prototype.clearTimer = function clearTimer(topic, uniqueID, cb) {
    var self = this;
    
    if(!self.ready) {
        debug('adding item to clearTimer queue');
        self.clearTimerQueue.push({
            topic: topic,
            uniqueID: uniqueID,
            cb: cb
        });
        
        return;
    }
    
    var timerQueue = self._timerQueueName(topic, uniqueID);
    
    debug(chalk.yellow('clearTimer'), timerQueue);

    self._timerChannel.deleteQueue(timerQueue, {}, function(err) {
        if(cb) cb(err);
    });
};


// rpc stuff
Rabbitr.prototype._rpcQueueName = function _rpcQueueName(topic) {
    return 'rpc.' + topic;
};
Rabbitr.prototype.rpcExec = function rpcExec(topic, d, opts, cb) {
    var self = this;
    
    if(!self.ready) {
        debug('adding item to rpcExec queue');
        self.rpcExecQueue.push({
            topic: topic,
            data: d,
            opts: opts,
            cb: cb
        });
        
        return;
    }

    if ('function' === typeof opts) {
        // shift arguments
        cb = opts;
        opts = {};
    }
    
    // this will send the data down the topic and then open up a unique return queue
    
    var rpcQueue = self._rpcQueueName(topic);
    
    var unique = shortId.generate() + '_' + ((Math.round(new Date().getTime() / 1000) + '').substr(5));
    var returnQueueName = rpcQueue + '.return.' + unique;
    
    var now = new Date().getTime();
        
    // bind the response queue
    var processed = false;

    self.connection.createChannel(function(err, channel) {
        if(err) {
            self.emit('error', err);
            if(cb) cb(err);
            return;
        }

        channel.assertQueue(self._formatName(returnQueueName), {
            exclusive: true,
            expires: (self.opts.defaultRPCExpiry*1 + 1000)
        });
        
        debug('using rpc return queue "%s"', chalk.cyan(returnQueueName));

        var cleanup = function() {
            // delete the return queue and close exc channel
            try {
                channel.deleteQueue(self._formatName(returnQueueName), function() {
                    channel.close();
                });
            }
            catch(e) {
                console.log('rabbitr cleanup exception', e);
            }
        };

        // set a timeout
        var timeoutMS = (opts.timeout || self.opts.defaultRPCExpiry || kDefaultRPCExpiry) * 1;
        var timeout = setTimeout(function() {
            debug('request timeout firing for', rpcQueue, 'to', returnQueueName);    

            cb({
                topic: rpcQueue,
                message: 'Request timed out'
            });
            
            cleanup();
        }, timeoutMS);

        var processMessage = function processMessage(msg) {
            if(!msg) return;

            var data = msg.content.toString();
            if(msg.properties.contentType == 'application/json') {
                data = JSON.parse(data);
            }

            if(processed) {
                cleanup();
                return; 
            }
            processed = true;
            
            clearTimeout(timeout);

            var error = data.error;
            var response = data.response;

            if (error) {
                error = JSON.parse(error);
                if (data.isError) {
                    var err = new Error(error.message);
                    Object.keys(error).forEach(function(key) {
                        if (err[key] !== error[key]) {
                            err[key] = error[key];
                        }
                    });
                    error = err;
                }
            }

            cb(error, response);
            
            cleanup();
        };
        channel.consume(self._formatName(returnQueueName), processMessage, {
            noAck: true
        }, function(err) {
            if(err) {
                self.emit('error', err);
                if(cb) cb(err);
                return;
            }

            // send the request now
            var data = {
                d: d,
                returnQueue: self._formatName(returnQueueName),
                expiration: now+timeoutMS
            };
            self._publishChannel.sendToQueue(self._formatName(rpcQueue), new Buffer(JSON.stringify(data)), {
                contentType: 'application/json',
                expiration: timeoutMS
            });
        });
    });
};

Rabbitr.prototype.rpcListener = function rpcListener(topic, opts, executor, alreadyInQueue) {
    var self = this;
    
    if(alreadyInQueue != true) {
        debug('adding item to rpcListener queue');
        self.rpcListenerQueue.push({
            topic: topic,
            executor: executor,
            opts: opts
        });
    }
    
    if ('function' === typeof opts) {
        // shift arguments
        alreadyInQueue = executor;
        executor = opts;
        opts = {};
    }

    if(!self.ready) {
        return;
    }
    
    var rpcQueue = self._rpcQueueName(topic);  

    opts.skipMiddleware = true;
    self.subscribe(rpcQueue, opts, function() { });
    
    debug('has rpcListener for', topic);
    
    self.on(rpcQueue, function(message) {
        var dataEnvelope = message.data;
        message.data = dataEnvelope.d;
        
        var now = new Date().getTime();
        
        if(now > dataEnvelope.expiration) {
            message.ack();
            
            return;
        }

        // support for older clients - is this needed?
        message.queue = {
            shift: message.ack
        };

        self._runMiddleware(message, function(middlewareErr) {
            // TODO - how to handle common error function thing for middleware?

            var _cb = function(err, response) {
                if(err) {
                    debug(chalk.cyan('rpcListener') + ' ' + chalk.red('hit error'), err);
                }

                // ack here - this will get ignored if the executor has acked or nacked already anyway
                message.ack();

                var isError = err instanceof Error;
                var errJSON = isError ? JSON.stringify(err, Object.keys(err).concat(['name', 'type', 'arguments', 'stack', 'message'])) : JSON.stringify(err);
                
                // doesn't need wrapping in self.formatName as the rpcExec function already formats the return queue name as required
                self._publishChannel.sendToQueue(dataEnvelope.returnQueue, new Buffer(JSON.stringify({
                    error: err ? errJSON : undefined,
                    isError: isError,
                    response: response,
                })), {
                    contentType: 'application/json'
                });
            };

            var _runExecutor = function() {
                executor(message, _cb);
            };

            if(!opts.middleware || opts.middleware.length === 0) {
                // no middleware specified for this listener
                return _runExecutor();
            }

            async.eachSeries(opts.middleware, function(middlewareFunc, next) {
                middlewareFunc(message, _cb, next);
            }, _runExecutor);
        });
    });
};

// message middleware support
Rabbitr.prototype.use = function use(middlewareFunc) {
    var self = this;
    
    self.middleware.push(middlewareFunc);
};
Rabbitr.prototype._runMiddleware = function _runMiddleware(message, next) {
    var self = this;

    if(self.middleware.length === 0) return next();

    async.eachSeries(self.middleware, function(middlewareFunc, next) {
        middlewareFunc(message, next);
    }, next);
};
